File: cache/cache-common/src/main/java/com/example/UserController.java
package com.example;

import java.util.List;

import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/users")
@RequiredArgsConstructor
public class UserController {
    private final UserService userService;

    @GetMapping
    public List<User> findAll() {
        return userService.findAll();
    }

    @GetMapping("/{id}")
    public User findById(@PathVariable int id) {
        return userService.findById(id);
    }

    @DeleteMapping
    public void evictCache() {
        userService.evictCache();
    }

    @DeleteMapping("/{id}")
    public void evictCache(@PathVariable int id) {
        userService.evictCache(id);
    }

    @PutMapping("/{id}")
    public User updateCache(@PathVariable int id) {
        return userService.updateCache(id);
    }
}

File: others/scala/src/main/scala/com/example/web/UserController.scala
package com.example.web

import java.net.URI

import com.example.entity.User
import com.example.service.UserService
import org.springframework.http.{HttpStatus, ResponseEntity}
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation._

import scala.jdk.CollectionConverters._

@RestController
@RequestMapping(Array("/users"))
class UserController(val userService: UserService) {

  @GetMapping
  def findAll(): java.lang.Iterable[User] = userService.findAll().asJava

  @GetMapping(params = Array("name"))
  def findByName(@RequestParam name: String): java.lang.Iterable[User] = userService.findByName(name).asJava

  @GetMapping(Array("/{id}"))
  def getOne(@PathVariable id: Int): User = userService.getOne(id)

  @PostMapping
  @ResponseStatus(HttpStatus.CREATED)
  def create(@RequestBody @Validated request: UserRequest): ResponseEntity[User] = {
    val user = new User(name = request.getName, age = request.getAge)
    val created = userService.create(user)
    ResponseEntity.created(URI.create("/api/users/" + created.getId)).body(created)
  }

  @PutMapping(Array("/{id}"))
  def update(@PathVariable id: Int, @RequestBody @Validated request: UserRequest): User = {
    val user = new User(id = id, name = request.getName, age = request.getAge)
    userService.update(user)
  }

  @DeleteMapping(Array("/{id}"))
  @ResponseStatus(HttpStatus.NO_CONTENT)
  def delete(@PathVariable id: Int): Unit = userService.delete(id)
}
